### PR TITLE
Add networking to XMTP client and expose it in FFI crate

### PR DESCRIPTION
**Pull request by richardhuaaa**
Note that half of the changes are in `bindings_ffi/src/lib.rs`, which gets automatically hidden as a large diff.

This is the starting point for exposing our libxmtp client to the platform layer (iOS/Android). I added the networking trait from @neekolas's PR with some slight modifications and hooked up `xmtp_networking` to it.

There's still a lot of gaps to fill here but figured I should land this first so there's some initial structure:
1. Have not tested whether this actually works with `swift-bridge` at the platform layer. I will switch to Mozilla uniffi first and test the integration at that point.
2. Need to add proper error handling.
3. Have not exposed account creation to the platform layer yet.
4. Have not hooked up the API auth token.

I will work on (1) next - if anyone wants to pull any of the others feel free!

### CODE REVIEW

File: xmtp/src/builder.rs [hunk 1]
```diff
@@ -56,8 +66,8 @@ where
         self
     }
 
-    pub fn wallet_address(mut self, wallet_address: String) -> Self {
-        self.wallet_address = Some(wallet_address);
+    pub fn wallet_address(mut self, wallet_address: &str) -> Self {
```
Comment (neekolas) [inline xmtp/src/builder.rs:69]: Curious why this change was necessary, given it's being converted back to a string

File: xmtp/src/client.rs [hunk 2]
```diff
@@ -11,16 +12,18 @@ pub enum Network {
     Prod,
 }
 
-pub struct Client<P>
+pub struct Client<A, P>
 where
+    A: XmtpApiClient,
     P: Persistence,
 {
+    pub api_client: A,
```
Comment (neekolas) [inline xmtp/src/client.rs:20]: Would be good to have a test that exercises this

**Review by neekolas — APPROVED** [review]
Great work! This unblocks a whole lot of workstreams

File: xmtp/src/builder.rs [hunk 3]
```diff
@@ -91,7 +101,13 @@ where
         }
     }
 
-    pub fn build(&mut self) -> Result<Client<P>, ClientBuilderError<P::Error>> {
+    pub fn build(&mut self) -> Result<Client<A, P>, ClientBuilderError<P::Error>> {
+        let api_client =
+            self.api_client
+                .take()
```
Comment (jazzz) [inline xmtp/src/builder.rs:107]: Taking the value of the Option leaves `self` in a changed state -- and multiple calls will produce different configurations of Clients. `build` should consume self if its going to consume it's members, to avoid confusion

File: bindings_ffi/src/lib.rs [hunk 4]
```diff
@@ -337,25 +158,28 @@ mod tests {
...
         let topic = Uuid::new_v4();
-        let mut client = super::create_client(ADDRESS.to_string(), false)
+        let mut client = super::create_client(account, ADDRESS, false).await.unwrap();
+        let mut sub = client
+            .api_client
+            .subscribe(vec![topic.to_string()])
             .await
             .unwrap();
-        let mut sub = client.subscribe(vec![topic.to_string()]).await.unwrap();
         std::thread::sleep(std::time::Duration::from_millis(100));
```
[the remark this answers is not in the document]
Comment (neekolas) [reply bindings_ffi/src/lib.rs:169]: Good catch. I've run into this one before too

File: xmtp/src/networking.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+#[async_trait]
+impl XmtpApiClient for MockXmtpApiClient {
+    async fn publish(
+        &mut self,
+        token: String,
+        envelopes: Vec<Envelope>,
+    ) -> Result<PublishResponse, String> {
+        let mut existing: Vec<Envelope> = match self.messages.get(&token) {
```
Comment (neekolas) [inline xmtp/src/networking.rs:54]: This is my half-finished garbage code. It shouldn't be looking up the messages by token, but going through each envelope and finding them by topic. You can see a correctly working version [here](https://github.com/xmtp/libxmtp/pull/118/files#diff-1c0e183cafb21ff8ca78b4d5a769750eb7df51979ca1f3316bd9a32786c9e645R54-R62).
